File: api-gateway/gateway/controllers/network.py
"""Network controller: wired configuration and Wi-Fi scan/connect/forget via
the os hardware agent."""
import logging

import grpc
from flask import Response, request

from .. import hardware
from ..helpers import _json, _publish_if_success
from . import api_bp

logger = logging.getLogger(__name__)


def _network_agent_error(exc: grpc.RpcError) -> Response:
    """Map a hardware-agent gRPC error to a 503 JSON Response."""
    detail = exc.details() if hasattr(exc, "details") else str(exc)
    logger.error("hardware agent RPC failed: %s", detail)
    return _json({"error": f"Hardware agent unavailable: {detail}"}, 503)


@api_bp.route('/api/v1/network/config', methods=['GET'])
def get_network_config():
    """GET /api/v1/network/config — gateway relay."""
    try:
        return _json(hardware.get_network_config())
    except grpc.RpcError as exc:
        return _network_agent_error(exc)
    except Exception as exc:  # noqa: BLE001
        return _json({"error": str(exc)}, 500)


@api_bp.route('/api/v1/network/config', methods=['POST'])
def set_network_config():
    """POST /api/v1/network/config — gateway relay."""
    body = request.get_json(silent=True) or {}
    method = body.get("method")
    if not method:
        return _json({"error": "'method' field is required"}, 400)
    try:
        result = hardware.set_network_config(
            interface=body.get("interface", "wired"), method=method,
            address=body.get("address"), prefix=body.get("prefix"),
            gateway=body.get("gateway"), dns=body.get("dns"))
    except grpc.RpcError as exc:
        return _network_agent_error(exc)
    except Exception as exc:  # noqa: BLE001
        return _json({"error": str(exc)}, 500)
    return _publish_if_success(_json(result), "network_config_changed", ["network"], data=result)


@api_bp.route('/api/v1/network/wifi/scan', methods=['GET'])
def scan_wifi():
    """GET /api/v1/network/wifi/scan — gateway relay."""
    try:
        return _json(hardware.scan_wifi())
    except grpc.RpcError as exc:
        return _network_agent_error(exc)
    except Exception as exc:  # noqa: BLE001
        return _json({"error": str(exc)}, 500)


@api_bp.route('/api/v1/network/wifi/connect', methods=['POST'])
def connect_wifi():
    """POST /api/v1/network/wifi/connect — gateway relay."""
    body = request.get_json(silent=True) or {}
    ssid = body.get("ssid")
    if not ssid:
        return _json({"error": "'ssid' field is required"}, 400)
    try:
        result = hardware.connect_wifi(ssid, body.get("password", ""))
    except grpc.RpcError as exc:
        return _network_agent_error(exc)
    except Exception as exc:  # noqa: BLE001
        return _json({"error": str(exc)}, 500)
    return _publish_if_success(_json(result), "network_config_changed", ["network"], data=result)


@api_bp.route('/api/v1/network/wifi/forget', methods=['POST'])
def forget_wifi():
    """POST /api/v1/network/wifi/forget — gateway relay."""
    body = request.get_json(silent=True) or {}
    ssid = body.get("ssid")
    if not ssid:
        return _json({"error": "'ssid' field is required"}, 400)
    try:
        result = hardware.forget_wifi(ssid)
    except grpc.RpcError as exc:
        return _network_agent_error(exc)
    except Exception as exc:  # noqa: BLE001
        return _json({"error": str(exc)}, 500)
    return _publish_if_success(_json(result), "network_config_changed", ["network"], data=result)

File: system-vision/src/api/wasm32/services/network.rs
//! Backend access layer (HTTP/SSE on wasm, Tauri IPC on native).

use crate::api::wasm32::http::fetch_api;

/// IPv4 configuration of a single managed interface (wired or wireless).
#[derive(Debug, Clone, serde::Deserialize)]
pub struct InterfaceConfig {
    #[serde(default)]
    pub name: String,
    #[serde(default = "default_method")]
    pub method: String,
    pub address: Option<String>,
    pub prefix: Option<u8>,
    pub gateway: Option<String>,
    #[serde(default)]
    pub dns: Vec<String>,
    #[serde(default)]
    pub present: bool,
}

/// Default method.
fn default_method() -> String {
    "auto".to_string()
}

/// Current Wi-Fi association state.
#[derive(Debug, Clone, Default, serde::Deserialize)]
pub struct WifiStatus {
    #[serde(default)]
    pub ssid: String,
    #[serde(default)]
    pub state: String,
    #[serde(default)]
    pub signal: i32,
}

/// Response from GET /api/v1/network/config (wired + Wi-Fi).
#[derive(Debug, Clone, serde::Deserialize)]
pub struct NetworkConfig {
    pub wired: InterfaceConfig,
    pub wifi: InterfaceConfig,
    #[serde(default)]
    pub wifi_status: WifiStatus,
}

/// Response from the IP-config POST endpoint.
#[derive(Debug, Clone, serde::Deserialize)]
pub struct NetworkSetResponse {
    pub success: bool,
    pub message: String,
}

/// A single Wi-Fi network from a scan.
#[derive(Debug, Clone, serde::Deserialize)]
pub struct WifiNetwork {
    pub ssid: String,
    #[serde(default)]
    pub signal: i32,
    #[serde(default)]
    pub security: String,
    #[serde(default)]
    pub in_use: bool,
    #[serde(default)]
    pub saved: bool,
}

/// A `WifiScanResponse` struct.
#[derive(Debug, Clone, serde::Deserialize)]
pub struct WifiScanResponse {
    #[serde(default)]
    pub networks: Vec<WifiNetwork>,
}

/// A `WifiConnectResponse` struct.
#[derive(Debug, Clone, serde::Deserialize)]
pub struct WifiConnectResponse {
    pub success: bool,
    #[serde(default)]
    pub state: String,
    pub message: String,
}

/// GET /api/v1/network/config
pub async fn get_network_config() -> Result<NetworkConfig, String> {
    fetch_api::<NetworkConfig>("/api/v1/network/config", "GET", None).await
}

/// POST /api/v1/network/config — applies an IPv4 configuration to `interface`
/// ("wired" | "wifi").
pub async fn set_network_config(
    interface: String,
    method: String,
    address: Option<String>,
    prefix: Option<u8>,
    gateway: Option<String>,
    dns: Vec<String>,
) -> Result<NetworkSetResponse, String> {
    let body = serde_json::json!({
        "interface": interface,
        "method": method,
        "address": address,
        "prefix": prefix,
        "gateway": gateway,
        "dns": dns,
    })
    .to_string();
    fetch_api::<NetworkSetResponse>("/api/v1/network/config", "POST", Some(&body)).await
}

/// GET /api/v1/network/wifi/scan — lists available Wi-Fi networks.
pub async fn scan_wifi() -> Result<WifiScanResponse, String> {
    fetch_api::<WifiScanResponse>("/api/v1/network/wifi/scan", "GET", None).await
}

/// POST /api/v1/network/wifi/connect — connects and verifies the password.
pub async fn connect_wifi(ssid: String, password: String) -> Result<WifiConnectResponse, String> {
    let body = serde_json::json!({ "ssid": ssid, "password": password }).to_string();
    fetch_api::<WifiConnectResponse>("/api/v1/network/wifi/connect", "POST", Some(&body)).await
}

/// POST /api/v1/network/wifi/forget — removes a saved Wi-Fi network.
pub async fn forget_wifi(ssid: String) -> Result<NetworkSetResponse, String> {
    let body = serde_json::json!({ "ssid": ssid }).to_string();
    fetch_api::<NetworkSetResponse>("/api/v1/network/wifi/forget", "POST", Some(&body)).await
}
